### DISCUSSION TITLE
How to access values from model states outside of the model itself?

### DISCUSSION THREAD

**Discussion by NickSdot** [spatie/laravel-model-states#173] [Q&A]
Hey guys,

first of all thanks for the nice package!

Am using multiple states for some models and everything works as expected. Now am in a situation where I need the color and other values of the state outside of the model itself. The purpose is, that I will group models in a visual way by state. 

Each visual group will have the state title, a status color and a description. All the data is available IN the model of course. But not outside of the foreach.

I've just created a plain PHP example of what I would like to achieve. Please note the comments. 

```php

// the main model, it hasMany App\Models\Findig
$report =  Report::find(1);

// here I group the relations of App\Models\Report by the state column
$groupedFindings = $report->findings->groupBy(function($item) {
    return $item->getAssessmentSlug();
});

// I loop thought the relations
foreach ($groupedFindings as  $findings)
{
    // here I would like to access functions of the State, like getColor();
    // but of course the state from the model is not available here.
    echo '<h1 id="' . $var->getSateColor() . '">' . $var->getSateColor() . '</h1>';

    // works as expected
    foreach($findings as $finding) {
        echo $finding->getIdentifier() . '<br/>';
        echo $finding->getTitle() . '<br/>';
        echo $finding->getLocation() . '<br/>';
        echo $finding->getDescription() . '<br/>';
    }
}
```

After spending endless time to figuring out whether there is a way to achieve this, I assume it's not or I am just not getting it right.

What I would need is something like `AuditFindingItem::getStates()` but with an parameter like `AuditFindingItem::getStates('withData')` which returns not only

```
Illuminate\Support\Collection {#1527 ▼
  #items: array:3 [▼
    "assessment" => Illuminate\Support\Collection {#1531 ▼
      #items: array:5 [▼
        0 => "critical"
        1 => "informational"
        2 => "major"
        3 => "medium"
        4 => "minor"
      ]
    }
    "status" => Illuminate\Support\Collection {#1529 ▼
      #items: array:6 [▼
        ...
      ]
    }
    "type" => Illuminate\Support\Collection {#1528 ▼
      #items: array:3 [▼
        ...
      ]
    }
  ]
}
```

but more like

```
  #items: array:3 [▼
    "assessment" => Illuminate\Support\Collection {#1531 ▼
      #items: array:5 [▼
        0 => "critical" [
                0 => function data
                1 => function data
                2 => function data
                3 => function data
                ]
        1 => "informational"
        2 => "major"
        3 => "medium"
        4 => "minor"
      ]
    }
```

I understand, that to create a class to get the data requires the specific model. 

But is there really no other way to access the data without being in the model context? Can anyone please point me in the right direction how I can achieve this, without moving most of the data out of the state class or maintaining twice?

Thanks in advance!

**Comment by brendt** [comment]
> But is there really no other way to access the data without being in the model context?

Nope, you could manually create a new state object and pass it a new model instance if you really want to.

**Reply by NickSdot** [reply]
Thanks for your answer! Mind to give me an idea how you would solve to get the data for the following purpose?

Let's say you want to create form selects with all the available states to be able to filter for it. Or you want to create a list of all models and group it by state. And in both cases you not only would need the the state name, but also some custom methods like the color. 

Would you really create for each of the state classes a 'dummy model' to access it?

Thanks in advance!

**Reply by brendt** [reply]
Yeah, we do something like this:

```php
[
	StateA::class => (new StateA(new MyModel))->getLabel(),
	StateB::class => (new StateB(new MyModel))->getLabel(),
	StateC::class => (new StateC(new MyModel))->getLabel(),
	StateD::class => (new StateD(new MyModel))->getLabel(),
]
```

But we automate it on the viewmodel level so that we can do something like this:

```php
public function stateOptions(): array
{
	return $this->stateOptions(MyModelState::class);
}
```
